File: typescript/packages/core/src/types/domain.ts
export type MessageRole = 'system' | 'user' | 'assistant' | 'tool';

export interface TextContent {
  type: 'text';
  text: string;
}

export interface ToolCallContent {
  type: 'tool_call';
  id: string;
  name: string;
  arguments: Record<string, unknown>;
}

export interface ToolResultContent {
  type: 'tool_result';
  toolCallId: string;
  content: string;
  isError?: boolean;
}

export type MessageContent = TextContent | ToolCallContent | ToolResultContent;

export interface Message {
  role: MessageRole;
  content: string | MessageContent[];
}

export interface UserMessage {
  role: 'user';
  content: string;
}

export interface TokenUsage {
  input: number;
  output: number;
  total: number;
}

export type AgentMode = 'agent' | 'plan' | 'ask';

export interface AgentRunRequest {
  sessionId: string;
  input: UserMessage;
  mode?: AgentMode;
  abortSignal?: AbortSignal;
}

export type RunStatus = 'pending' | 'running' | 'completed' | 'failed' | 'cancelled';

export type TerminalReason =
  | { kind: 'completed' }
  | { kind: 'max_turns' }
  | { kind: 'cancelled' }
  | { kind: 'error'; message: string };

export interface ToolExecutionContext {
  sessionId: string;
  runId: string;
  workspaceRoot: string;
  abortSignal?: AbortSignal;
}

export interface ToolDefinitionSchema {
  type: 'object';
  properties: Record<string, unknown>;
  required?: string[];
}

export type ToolConcurrency = 'parallel' | 'serial';

export interface ToolDefinition {
  name: string;
  description: string;
  parameters: ToolDefinitionSchema;
  concurrency?: ToolConcurrency;
}

export interface ToolCallRequest {
  id: string;
  name: string;
  arguments: Record<string, unknown>;
}

export interface ModelChunk {
  type: 'text_delta' | 'tool_call' | 'usage' | 'done';
  textDelta?: string;
  toolCall?: ToolCallRequest;
  usage?: TokenUsage;
}

export interface StreamRequest {
  messages: Message[];
  tools: ToolDefinition[];
  abortSignal?: AbortSignal;
}

export type PermissionDecision = 'allow' | 'deny' | 'ask';

export interface PermissionCheckRequest {
  toolName: string;
  arguments: Record<string, unknown>;
  sessionId: string;
}

export interface HookContext {
  sessionId: string;
  runId: string;
}

export type HookEvent =
  | { type: 'onRunStart'; context: HookContext }
  | { type: 'onRunEnd'; context: HookContext; reason: TerminalReason }
  | { type: 'preToolUse'; context: HookContext; toolName: string; arguments: Record<string, unknown> }
  | { type: 'postToolUse'; context: HookContext; toolName: string; result: unknown; error?: string };

export interface PluginManifest {
  id: string;
  version: string;
  name: string;
  capabilities: string[];
  configSchema?: Record<string, unknown>;
}

export interface AgentConfig {
  workspaceRoot: string;
  maxTurns: number;
  model: string;
  provider: string;
  toolTimeoutMs: number;
  runTimeoutMs: number;
  tokenBudget: number;
}

export const DEFAULT_AGENT_CONFIG: AgentConfig = {
  workspaceRoot: process.cwd(),
  maxTurns: 25,
  model: 'gpt-4o-mini',
  provider: 'openai',
  toolTimeoutMs: 60_000,
  runTimeoutMs: 600_000,
  tokenBudget: 100_000,
};

File: python/src/agent_framework/core/domain.py
from __future__ import annotations

import asyncio

from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Literal, TypedDict, Union

MessageRole = Literal["system", "user", "assistant", "tool"]
AgentMode = Literal["agent", "plan", "ask"]
RunStatus = Literal["pending", "running", "completed", "failed", "cancelled"]
ToolConcurrency = Literal["parallel", "serial"]
PermissionDecision = Literal["allow", "deny", "ask"]


class TokenUsage(TypedDict):
    input: int
    output: int
    total: int


class UserMessage(TypedDict):
    role: Literal["user"]
    content: str


@dataclass
class TextContent:
    type: Literal["text"] = "text"
    text: str = ""


@dataclass
class ToolCallContent:
    type: Literal["tool_call"] = "tool_call"
    id: str = ""
    name: str = ""
    arguments: dict[str, Any] = field(default_factory=dict)


@dataclass
class ToolResultContent:
    type: Literal["tool_result"] = "tool_result"
    tool_call_id: str = ""
    content: str = ""
    is_error: bool = False


MessageContent = Union[TextContent, ToolCallContent, ToolResultContent]


@dataclass
class Message:
    role: MessageRole
    content: str | list[MessageContent]


@dataclass
class AgentRunRequest:
    session_id: str
    input: UserMessage
    mode: AgentMode = "agent"
    cancel_event: asyncio.Event | None = None


@dataclass
class ToolDefinition:
    name: str
    description: str
    parameters: dict[str, Any]
    concurrency: ToolConcurrency = "parallel"


@dataclass
class ToolCallRequest:
    id: str
    name: str
    arguments: dict[str, Any]


@dataclass
class ModelChunk:
    type: Literal["text_delta", "tool_call", "usage", "done"]
    text_delta: str | None = None
    tool_call: ToolCallRequest | None = None
    usage: TokenUsage | None = None


@dataclass
class ToolExecutionContext:
    session_id: str
    run_id: str
    workspace_root: str


@dataclass
class AgentConfig:
    workspace_root: str = "."
    max_turns: int = 25
    model: str = "gpt-4o-mini"
    provider: str = "openai"
    tool_timeout_ms: int = 60_000
    run_timeout_ms: int = 600_000
    token_budget: int = 100_000


DEFAULT_AGENT_CONFIG = AgentConfig()


class AgentErrorCode(str, Enum):
    CANCELLED = "CANCELLED"
    PERMISSION_DENIED = "PERMISSION_DENIED"
    TOOL_NOT_FOUND = "TOOL_NOT_FOUND"
    TOOL_EXECUTION_FAILED = "TOOL_EXECUTION_FAILED"
    PROVIDER_ERROR = "PROVIDER_ERROR"
    MAX_TURNS = "MAX_TURNS"
    VALIDATION_ERROR = "VALIDATION_ERROR"
    TIMEOUT = "TIMEOUT"


class AgentError(Exception):
    def __init__(self, message: str, code: AgentErrorCode, recoverable: bool = False):
        super().__init__(message)
        self.code = code
        self.recoverable = recoverable


TerminalReason = dict[str, str]
